WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/applicationpassword/ApplicationPasswordViewModelSlice.kt

```diff
@@ -0,0 +1,153 @@
+
+    val uiModelMutable = MutableLiveData<MySiteCardAndItem.Card?>()
+    val uiModel: LiveData<MySiteCardAndItem.Card?> = uiModelMutable
+
+    @Suppress("UnusedParameter")
+    fun buildCard(siteModel: SiteModel) {
+        // This is hidden for regular users.
+        // After enabling it, please remove the Suppress annotation for buildCard and buildApplicationPasswordDiscovery
+        buildApplicationPasswordDiscovery(siteModel)
+    }
+
+    @Suppress("UnusedPrivateMember")
+    private fun buildApplicationPasswordDiscovery(site: SiteModel) {
+        // Check if the site URL is already cached
+        val cachedValue = siteURLCache[site.url]
+        if (cachedValue == null) {
+            // No cached value, set to null until get a response
+            uiModelMutable.postValue(null)
+        } else {
+            // If cached value is empty, it means the site has no authentication URL
+            if (cachedValue.isEmpty()) {
+                uiModelMutable.postValue(null)
+            } else {
+                postAuthenticationUrl(cachedValue)
+            }
+            return
+        }
+
+        scope.launch {
+            // If the site is already authorized, no need to run the discovery
+            val storedSite = siteSqlUtils.getSiteWithLocalId(site.localId())
+            if (storedSite != null &&
+                !storedSite.apiRestUsername.isNullOrEmpty() && !storedSite.apiRestPassword.isNullOrEmpty()) {
+                return@launch
+            }
+
+            val authorizationUrlComplete = getAuthorizationUrlComplete(site.url)
+            if (authorizationUrlComplete.isEmpty()) {
+                uiModelMutable.postValue(null)
+                siteURLCache[site.url] = ""
+            } else {
+                postAuthenticationUrl(authorizationUrlComplete)
+                siteURLCache[site.url] = authorizationUrlComplete
+            }
+        }
+    }
+
+    private fun postAuthenticationUrl(authorizationUrlComplete: String) {
+        uiModelMutable.postValue(
+            MySiteCardAndItem.Card.QuickLinksItem(
+                listOf(
+                    QuickLinkItem(
+                        label = UiString.UiStringRes(R.string.application_password_title),
+                        icon = R.drawable.ic_lock_white_24dp,
+                        onClick = onClick(authorizationUrlComplete)
+                    )
+                )
+            )
+        )
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    private suspend fun getAuthorizationUrlComplete(siteUrl: String): String = withContext(bgDispatcher) {
+        try {
+            getAuthorizationUrlCompleteInternal(siteUrl)
+        } catch (throwable: Throwable) {
+            handleAuthenticationDiscoveryError(siteUrl, throwable)
+        }
+    }
+
+    private fun handleAuthenticationDiscoveryError(siteUrl: String, throwable: Throwable): String {
+        Log.e("WP_RS", "VM: Error during API discovery for $siteUrl", throwable)
```

Comment:
I recommend using [AppLogWrapper](https://github.com/wordpress-mobile/WordPress-Android/blob/3cb93923319fbfba5f46e90da9017232c7c00a92/libs/fluxc/src/main/java/org/wordpress/android/fluxc/utils/AppLogWrapper.kt#L6) instead of `Log` here. It can be passed to the constructor as a private va, ie:

```
private val appLogWrapper: AppLogWrapper
```